### PR TITLE
Add add-on version list/detail API, with new fields (fix #3025)

### LINKED ISSUE #3025: Need an API to list all versions for an add-on
We need an API listing all versions for an add-on. 

In addition, we need to be able to display release notes and license information for each and link to the source if enabled (These new fields might not be needed in the add-on detail/search API, the detail page could call an extra API when displaying those, maybe).

Supports mozilla/addons/issues/151

### PR DESCRIPTION
**Pull request by diox**
Fix #3025

### CODE REVIEW

File: docs/topics/api/addons.rst [hunk 1]
```diff
@@ -18,12 +18,12 @@ This endpoint allows you to search through public add-ons.
...
-    :param string appversion: Filter by application version compatibility. Pass the full version as a string, e.g. ``46.0``. Only valid when the ``app`` parameter is also present.
-    :param string platform: Filter by :ref:`add-on platform <addon-detail-platform>` availability.
-    :param string type: Filter by :ref:`add-on type <addon-detail-type>`.
-    :param string sort: The sort parameter. The available parameters are documented in the :ref:`table below <addon-search-sort>`.
+    :query string q: The search query.
+    :query string app: Filter by :ref:`add-on application <addon-detail-application>` availability.
+    :query string appversion: Filter by application version compatibility. Pass the full version as a string, e.g. ``46.0``. Only valid when the ``app`` parameter is also present.
+    :query string platform: Filter by :ref:`add-on platform <addon-detail-platform>` availability.
+    :query string type: Filter by :ref:`add-on type <addon-detail-type>`.
+    :query string sort: The sort parameter. The available parameters are documented in the :ref:`table below <addon-search-sort>`.
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Good catch, `:query:` is certainly more correct here.

File: docs/topics/api/addons.rst [hunk 2]
```diff
@@ -192,6 +180,76 @@ This endpoint allows you to fetch a specific add-on by id, slug or guid.
         dictionary  Dictionary
     ==============  ==========================================================
 
+-------------
+Versions List
+-------------
+
+.. _version-list:
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Should this be ".. _version**s**-list"?

File: docs/topics/api/addons.rst [hunk 3]
```diff
@@ -192,6 +180,76 @@ This endpoint allows you to fetch a specific add-on by id, slug or guid.
...
+
+This endpoint allows you to list all versions belonging to a specific add-on.
+
+.. http:get:: /api/v3/addons/addon/(int:addon_id|string:addon_slug|string:addon_guid)/versions/
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: to be honest I feel like at some point we should embrace restricting and using the `addon_guid` as the one true source of add-on identification if that's possible.

File: docs/topics/api/addons.rst [hunk 4]
```diff
@@ -192,6 +180,76 @@ This endpoint allows you to fetch a specific add-on by id, slug or guid.
...
+        reviewer permissions or a user account listed as a developer of the
+        add-on.
+
+    :query string filter: The :ref:`filter <version-filtering-param>` to apply.
+    :>json int count: The number of versions for this add-on.
+    :>json string next: The URL of the next page of results.
+    :>json string previous: The URL of the previous page of results.
+    :>json array results: An array of :ref:`versions <version-detail-object>`.
+
+.. _version-filtering-param:
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Also version**s**?

File: docs/topics/api/addons.rst [hunk 5]
```diff
@@ -192,6 +180,76 @@ This endpoint allows you to fetch a specific add-on by id, slug or guid.
...
+
+   By default, the version list API will only return versions with valid statuses
+   (excluding versions that have incomplete, disabled, deleted, rejected or
+   flagged for further review files) - you can change that with the ``filter``
+   query parameter, which requires authentication and specific rights depending
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: maybe "permissions" is a better, more exact word than "rights"?

File: docs/topics/api/addons.rst (hunk 3 above)
Comment (diox) [inline docs/topics/api/addons.rst]: I find the choice between the 3 identifiers very useful for quickly access information about an add-on, so I dunno. The add-on detail page at least need to have that choice to allow admins to find an add-on with one of those identifiers.

File: docs/topics/api/addons.rst [hunk 6]
```diff
@@ -192,6 +180,76 @@ This endpoint allows you to fetch a specific add-on by id, slug or guid.
...
+    :>json array files: Array holding information about the files for the version.
+    :>json int files[].id: The id for a file.
+    :>json string files[].created: The creation date for a file.
+    :>json string files[].hash: The hash for a file.
+    :>json string files[].platform: The :ref:`platform <addon-detail-platform>` for a file.
+    :>json int files[].id: The size for a file, in bytes.
+    :>json int files[].status: The :ref:`status <addon-detail-status>` for a file.
+    :>json string files[].url: The (absolute) URL to download a file.
+    :>json object license: Object holding information about the license for the version.
+    :>json string|object|null license.name: The name of the license.
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: What kind of `object` is allowed here? Is that for translations?

Comment (diox) [inline docs/topics/api/addons.rst]: Right. Works as described in http://addons-server.readthedocs.io/en/latest/topics/api/overview.html#responses

File: src/olympia/addons/views.py [hunk 7]
```diff
@@ -644,6 +645,63 @@ def feature_compatibility(self, request, pk=None):
...
+            self.queryset = Version.unfiltered.all()
+        # When accessing a single version or if requesting it explicitly when
+        # listing, reviewers and add-on authors can access all non-deleted
+        # versions.
+        elif ((requested == 'all' or self.action != 'list') and (
+            AllowReviewer().has_permission(self.request, self) or
+                AllowAddonAuthor().has_object_permission(
+                    self.request, self, self.get_addon_object()))):
+            self.queryset = Version.objects.all()
+
```
Comment (EnTeQuAk) [inline src/olympia/addons/views.py]: I'm still a fan of doing something like

```python
can_access_unfiltered_versions = (
    (requested == 'all_with_deleted' or self.action != 'list') and
    # ...
)

can_access_filtered_versions = (
    ...
)

if can_accecss_unfiltered_versions:
    # ...
elif can_access_filtered_versions:
    # ...
```

or something similar compared to bloating the actual `if` statements. It's a minor nitpick though and nothing that has to be changed.

File: src/olympia/addons/views.py [hunk 8]
```diff
@@ -644,6 +645,63 @@ def feature_compatibility(self, request, pk=None):
...
+        # Now that the base queryset has been altered, call super() to use it.
+        qs = super(AddonVersionViewSet, self).get_queryset()
+        # Filter with the add-on.
+        return qs.filter(
+            addon=self.get_addon_object())
```
Comment (EnTeQuAk) [inline src/olympia/addons/views.py]: The statement looks so short, does that need breaking the line?

File: docs/topics/api/addons.rst (hunk 3 above)
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Hmm, guess you're right, acccessing information quickly is a valid point too.

File: docs/topics/api/addons.rst (hunk 6 above)
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Right, maybe we should or maybe we shouldn't link to that page somehow or at least note in the description of those fields that there can be translations.

**Comment by EnTeQuAk** [comment]
r+wc

File: src/olympia/addons/views.py (hunk 8 above)
Comment (diox) [inline src/olympia/addons/views.py]: Probably not, I had something else in there initially and forgot to clean it up

File: src/olympia/addons/views.py (hunk 7 above)
Comment (diox) [inline src/olympia/addons/views.py]: I'll try that

File: docs/topics/api/addons.rst (hunk 6 above)
Comment (diox) [inline docs/topics/api/addons.rst]: Yeah. I'll go back to the doc and add a link to the translations overview for each translation.

Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Awesome, that'll help.

**Pull request merged**